### PR TITLE
[SPARK-6940][MLLIB] Add CrossValidator to Python ML pipeline API

### PR DESCRIPTION
**Pull request by mengxr**
Since CrossValidator is a meta algorithm, we copy the implementation in Python. @jkbradley

### CODE REVIEW

File: python/pyspark/ml/pipeline.py [hunk 1]
```diff
@@ -71,6 +71,15 @@ def transform(self, dataset, params={}):
 
 
 @inherit_doc
+class Model(Transformer):
+    """
+    Abstract class for models that fitted by estimators.
```
Comment (jkbradley) [inline python/pyspark/ml/pipeline.py]: "that fitted" --> "that are fitted"

File: python/pyspark/ml/tuning.py [hunk 2]
```diff
@@ -79,6 +85,173 @@ def build(self):
...
+    """
+
+    # a placeholder to make it appear in the generated doc
+    estimator = Param(Params._dummy(), "estimator", "estimator to be cross-validated")
+
+    # a placeholder to make it appear in the generated doc
+    estimatorParamMaps = Param(Params._dummy(), "estimatorParamMaps", "estimator param maps")
+
+    # a placeholder to make it appear in the generated doc
+    evaluator = Param(Params._dummy(), "evaluator", "evaluator for selection")
```
Comment (jkbradley) [inline python/pyspark/ml/tuning.py]: Here and in Scala, we should say CrossValidator tries to maximize the evaluation metric.

File: python/pyspark/ml/tuning.py [hunk 3]
```diff
@@ -79,6 +85,173 @@ def build(self):
...
+        self.paramMap[self.numFolds] = value
+        return self
+
+    def getNumFolds(self):
+        """
+        Gets the value of numFolds or its default value.
+        """
+        return self.getOrDefault(self.numFolds)
+
+    def fit(self, dataset, params={}):
```
Comment (jkbradley) [inline python/pyspark/ml/tuning.py]: Here and in Scala, we should check to see if users pass in parameters for the evaluator.

File: python/pyspark/ml/tuning.py [hunk 4]
```diff
@@ -79,6 +85,173 @@ def build(self):
...
+        paramMap = self.extractParamMap(params)
+        est = paramMap[self.estimator]
+        epm = paramMap[self.estimatorParamMaps]
+        numModels = len(epm)
+        eva = paramMap[self.evaluator]
+        nFolds = paramMap[self.numFolds]
+        h = 1.0 / nFolds
```
Comment (jkbradley) [inline python/pyspark/ml/tuning.py]: Until we have per-parameter validation, we should check for nFolds < 2 here

File: python/pyspark/ml/tuning.py [hunk 5]
```diff
@@ -79,6 +85,173 @@ def build(self):
...
+                metric = eva.evaluate(model.transform(validation, epm[j]))
+                metrics[j] += metric
+        bestIndex = np.argmax(metrics)
+        bestModel = est.fit(dataset, epm[bestIndex])
+        return CrossValidatorModel(bestModel)
+
+
+class CrossValidatorModel(Model):
+    """
+    Model from k-fold corss validation.
```
Comment (jkbradley) [inline python/pyspark/ml/tuning.py]: typo: "corss"

**Comment by jkbradley** [comment]
@mengxr LGTM other than the small comments above

File: python/pyspark/ml/tuning.py (hunk 3 above)
Comment (mengxr) [inline python/pyspark/ml/tuning.py]: This could be done by implementing the `copy` method (in a separate PR).

Comment (jkbradley) [inline python/pyspark/ml/tuning.py]: Good point; I guess I'm doing that right now for [https://issues.apache.org/jira/browse/SPARK-7380]

**Comment by jkbradley** [comment]
LGTM

**Comment by mengxr** [comment]
Merged into master and branch-1.4.

**Pull request closed without merging**